packages/gestalt/src/ComboBox.js

```diff
@@ -151,87 +160,164 @@ const ComboBoxWithForwardRef: React$AbstractComponent<Props, HTMLInputElement> =
+        handleKeyNavigation(event, KEYS.DOWN);
+        event.preventDefault();
+      } else if (keyCode === ENTER) {
+        handleKeyNavigation(event, KEYS.ENTER);
+        event.stopPropagation();
+      } else if (keyCode === ESCAPE) {
+        if (innerRef) innerRef.current?.focus();
+      } else if (keyCode === TAB) {
+        setShowOptionsList(false);
+      }
+    },
+    [handleKeyNavigation],
+  );
 
-    setHoveredItemIndex(cursorIndex);
+  const handleOnDismiss = useCallback(() => setShowOptionsList(false), []);
 
-    const optionItem = suggestedOptions[cursorIndex];
+  // ==== EVENT HANDLING: InternalTextField ====
 
-    if (optionItem && direction === KEYS.ENTER) {
-      handleSelectOptionItem({ event, item: optionItem });
-    }
-  };
-
-  const handleKeyDown = (event) => {
-    const { keyCode } = event;
-
-    if (keyCode === UP_ARROW) {
-      handleKeyNavigation(event, KEYS.UP);
-      event.preventDefault();
-    } else if (keyCode === DOWN_ARROW) {
-      handleKeyNavigation(event, KEYS.DOWN);
-      event.preventDefault();
-    } else if (keyCode === ENTER) {
-      handleKeyNavigation(event, KEYS.ENTER);
-      event.stopPropagation();
-    } else if (keyCode === ESCAPE) {
-      if (innerRef) innerRef.current?.focus();
-    } else if (keyCode === TAB) {
-      setShowOptionsList(false);
+  const handleOnBlur = useCallback(({ event, value }) => onBlur?.({ event, value }), [onBlur]);
+
+  const handleOnFocus = useCallback(({ event, value }) => onFocus?.({ event, value }), [onFocus]);
+
+  const handleSetShowOptionsList = useCallback(() => setShowOptionsList(true), []);
+
+  const handleOnChange = useCallback(
+    ({ event, value }) => {
+      setHoveredItemIndex(null);
+      if (isNotControlled) {
+        setSelectedItem(null);
+        setTextfieldInput(value);
+      }
+      if (showOptionsList === false) setShowOptionsList(true);
+      onChange?.({ event, value });
+    },
+    [isNotControlled, onChange, showOptionsList],
+  );
+
+  const handleOnClickIconButtonClear = useCallback(() => {
+    setHoveredItemIndex(null);
+    if (isNotControlled) {
+      setSelectedItem(null);
+      setTextfieldInput('');
+      setSuggestedOptions(options);
     }
-  };
+    onClear?.();
+    innerRef?.current?.focus();
+  }, [isNotControlled, onClear, options]);
+
+  const handleOnKeyDown = useCallback(
+    ({ event, value }) => {
+      if (!showOptionsList && event.keyCode !== TAB) setShowOptionsList(true);
+      onKeyDown?.({ event, value });
+    },
+    [onKeyDown, showOptionsList],
+  );
 
-  const textfieldIconButton =
-    (controlledInputValue && controlledInputValue !== '') ||
-    (textfieldInput && textfieldInput !== '') ||
-    (tags && tags.length > 0)
-      ? 'clear'
-      : 'expand';
+  // ==== MAPPING ComboBoxItem ====
+  const comboBoxItemList = useMemo(() => {
+    const buildComboBoxItemCallback = ({ label: comboBoxItemlabel, subtext, value }, index) => {
```

Comment:
I think this might be more readable if you inline this into the `.map` call

Reply:
DONE